File: Cache_memory/cpp/matrix_block.cpp
#include <iostream>
#include <chrono>
#include <fstream>
#include <vector>

using namespace std;

void run_experiment(int N, int BLOCK_SIZE, ofstream &csv_file) {
    vector<vector<double>> A(N, vector<double>(N));
    vector<vector<double>> B(N, vector<double>(N));
    vector<vector<double>> C(N, vector<double>(N, 0));

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            A[i][j] = (i + j) * 0.5;
            B[i][j] = (i - j) * 0.3;
        }
    }

    auto start = chrono::high_resolution_clock::now();

    for (int bi = 0; bi < N; bi += BLOCK_SIZE) {
        for (int bj = 0; bj < N; bj += BLOCK_SIZE) {
            for (int bk = 0; bk < N; bk += BLOCK_SIZE) {
                for (int i = bi; i < min(bi + BLOCK_SIZE, N); i++) {
                    for (int j = bj; j < min(bj + BLOCK_SIZE, N); j++) {
                        for (int k = bk; k < min(bk + BLOCK_SIZE, N); k++) {
                            C[i][j] += A[i][k] * B[k][j];
                        }
                    }
                }
            }
        }
    }

    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> elapsed = end - start;

    csv_file << N << "," << BLOCK_SIZE << "," << elapsed.count() << endl;

    cout << "N = " << N << ", BLOCK_SIZE = " << BLOCK_SIZE << " | Tiempo: " << elapsed.count() << " s" << endl;
}

int main() {
    ofstream csv_file("matrix_block_results.csv");
    csv_file << "Dimension,Tamaño de Bloque,Tiempo\n";

    int sizes[] = {10, 50, 100, 200, 500};
    int block_sizes[] = {16};

    for (int N : sizes) {
        for (int BLOCK_SIZE : block_sizes) {
            run_experiment(N, BLOCK_SIZE, csv_file);
        }
    }

    csv_file.close();
    cout << "Resultados guardados en 'matrix_block_results.csv'\n";
    return 0;
}

File: Cache_memory/go/matrix_block.go
package main

import (
	"fmt"
	"time"
)

const BLOCK_SIZE = 32

func multiplyMatrices(N int) time.Duration {
	A := make([][]float64, N)
	B := make([][]float64, N)
	C := make([][]float64, N)
	for i := range A {
		A[i] = make([]float64, N)
		B[i] = make([]float64, N)
		C[i] = make([]float64, N)
	}

	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			A[i][j] = float64(i+j) * 0.5
			B[i][j] = float64(i-j) * 0.3
		}
	}

	start := time.Now()

	for bi := 0; bi < N; bi += BLOCK_SIZE {
		for bj := 0; bj < N; bj += BLOCK_SIZE {
			for bk := 0; bk < N; bk += BLOCK_SIZE {
				// Multiplicación dentro de los bloques
				for i := bi; i < min(bi+BLOCK_SIZE, N); i++ {
					for j := bj; j < min(bj+BLOCK_SIZE, N); j++ {
						for k := bk; k < min(bk+BLOCK_SIZE, N); k++ {
							C[i][j] += A[i][k] * B[k][j]
						}
					}
				}
			}
		}
	}

	return time.Since(start)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	dimensions := []int{10, 50, 100, 200, 500}

	for _, N := range dimensions {
		elapsed := multiplyMatrices(N)
		fmt.Printf("N = %d, BLOCK_SIZE = %d | Tiempo: %v\n", N, BLOCK_SIZE, elapsed)
	}
}
